store/tikv/tikvrpc/tikvrpc.go

```diff
@@ -0,0 +1,159 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tikvrpc
+
+import (
+	"fmt"
+
+	"github.com/pingcap/kvproto/pkg/coprocessor"
+	"github.com/pingcap/kvproto/pkg/errorpb"
+	"github.com/pingcap/kvproto/pkg/kvrpcpb"
+)
+
+// CmdType represents the concrete request type in Request or response type in Response.
+type CmdType uint16
+
+// CmdType values.
+const (
+	CmdGet CmdType = 1 + iota
+	CmdScan
+	CmdPrewrite
+	CmdCommit
+	CmdCleanup
+	CmdBatchGet
+	CmdBatchRollback
+	CmdScanLock
+	CmdResolveLock
+	CmdGC
+
+	CmdRawGet CmdType = 256 + iota
+	CmdRawPut
+	CmdRawDelete
+
+	CmdCop CmdType = 512 + iota
+)
+
+// Request wraps all kv/coprocessor requests.
+type Request struct {
+	Type          CmdType
+	Get           *kvrpcpb.GetRequest
+	Scan          *kvrpcpb.ScanRequest
+	Prewrite      *kvrpcpb.PrewriteRequest
+	Commit        *kvrpcpb.CommitRequest
+	Cleanup       *kvrpcpb.CleanupRequest
+	BatchGet      *kvrpcpb.BatchGetRequest
+	BatchRollback *kvrpcpb.BatchRollbackRequest
+	ScanLock      *kvrpcpb.ScanLockRequest
+	ResolveLock   *kvrpcpb.ResolveLockRequest
+	GC            *kvrpcpb.GCRequest
+	RawGet        *kvrpcpb.RawGetRequest
+	RawPut        *kvrpcpb.RawPutRequest
+	RawDelete     *kvrpcpb.RawDeleteRequest
+	Cop           *coprocessor.Request
+}
+
+// GetContext returns the rpc context for the underlying concrete request.
+func (req *Request) GetContext() (*kvrpcpb.Context, error) {
+	var c *kvrpcpb.Context
+	switch req.Type {
+	case CmdGet:
+		c = req.Get.GetContext()
+	case CmdScan:
+		c = req.Scan.GetContext()
+	case CmdPrewrite:
+		c = req.Prewrite.GetContext()
+	case CmdCommit:
+		c = req.Commit.GetContext()
+	case CmdCleanup:
+		c = req.Cleanup.GetContext()
+	case CmdBatchGet:
+		c = req.BatchGet.GetContext()
+	case CmdBatchRollback:
+		c = req.BatchRollback.GetContext()
+	case CmdScanLock:
+		c = req.ScanLock.GetContext()
+	case CmdResolveLock:
+		c = req.ResolveLock.GetContext()
+	case CmdGC:
+		c = req.GC.GetContext()
+	case CmdRawGet:
+		c = req.RawGet.GetContext()
+	case CmdRawPut:
+		c = req.RawPut.GetContext()
+	case CmdRawDelete:
+		c = req.RawDelete.GetContext()
+	case CmdCop:
+		c = req.Cop.GetContext()
+	default:
+		return nil, fmt.Errorf("invalid request type %v", req.Type)
+	}
+	return c, nil
+}
+
+// Response wraps all kv/coprocessor responses.
+type Response struct {
+	Type          CmdType
+	Get           *kvrpcpb.GetResponse
+	Scan          *kvrpcpb.ScanResponse
+	Prewrite      *kvrpcpb.PrewriteResponse
+	Commit        *kvrpcpb.CommitResponse
+	Cleanup       *kvrpcpb.CleanupResponse
+	BatchGet      *kvrpcpb.BatchGetResponse
+	BatchRollback *kvrpcpb.BatchRollbackResponse
+	ScanLock      *kvrpcpb.ScanLockResponse
+	ResolveLock   *kvrpcpb.ResolveLockResponse
+	GC            *kvrpcpb.GCResponse
+	RawGet        *kvrpcpb.RawGetResponse
+	RawPut        *kvrpcpb.RawPutResponse
+	RawDelete     *kvrpcpb.RawDeleteResponse
+	Cop           *coprocessor.Response
```

Comment:
It needs to define a struct to hold all the concrete response type here, just like the old kvproto definition of `msgpb.Response`(to make the change of code minimum, it just does the mimesis to old code). So the listing is inevitable, `tikvrpc.Response` should be defined as it's now, and the reflection is not necessary.
Since the `tikvrpc.Response` is used as a simple struct holder directly, it's not helpful to add a interface definition of `GetRegionError()`.